service/storage_service.cc: Make effective_ownership preemptable

amnonh opened the pull request:
A lot is going on when calculating effective ownership.
For each node in the cluster, we need to go over all the ranges belong
to that node and see if that node is the owner or not.

This patch uses futurized loops with do_for_each so it would preempt if
needed.

The patch replaces the current for-loops with do_for_each and do_with
but does not replace the logic.

Fixes #6380

avikivity commented on service/storage_service.cc:
```diff
@@ -1876,33 +1876,56 @@ future<std::map<gms::inet_address, float>> storage_service::effective_ownership(
-        }
-        return final_ownership;
+
+        auto final_ownership = make_lw_shared<std::map<gms::inet_address, float>>();
```
Why not put final_ownership in the  do_with? you can return it from within the do_with().

amnonh replied:
I had a poor experience with returning values own by do_with in the past, I can try

avikivity commented on service/storage_service.cc:
```diff
@@ -1876,33 +1876,56 @@ future<std::map<gms::inet_address, float>> storage_service::effective_ownership(
+        // DC and all the instances in each DC.
+        //
+        // The call for get_range_for_endpoint is done once per endpoint
+        return do_with(std::map<token, float>(dht::token::describe_ownership(ss._token_metadata.sorted_tokens())),
```
Why the extra constructor call? describe_ownership already returns a map<token, float>.

amnonh replied:
I'll add a move

avikivity replied:
Why?

amnonh replied:
To remove the constructor call

avikivity replied:
It's enough to remove it. No need for an extra move.

amnonh replied:
if you are ok with not explicitly stating the type, great

avikivity replied:
I don't understand.

amnonh replied:
I meant that type will be deduce and will not be stated explicitly.

avikivity replied:
The type is explicit in the argument list of the lambda:

```c++
        [&ss, keyspace_name, final_ownership](const std::map<token, float>& token_ownership, 
```

avikivity commented on service/storage_service.cc:
```diff
@@ -1876,33 +1876,56 @@ future<std::map<gms::inet_address, float>> storage_service::effective_ownership(
+                    return do_for_each(endpoints_map, [&ss, &keyspace_name, final_ownership, &token_ownership, &endpoints](const gms::inet_address& endpoint) {
+                        // calculate the ownership with replication and add the endpoint to the final ownership map
+                        try {
+                            return do_with(float(0.0f), dht::token_range_vector(ss.get_ranges_for_endpoint(keyspace_name, endpoint)),
```
get_ranges_for_endpoint() is also a nested loop!

amnonh replied:
it is, but not a too bad one, assuming that we are preempting between each call, I belive it should be ok

avikivity commented on service/storage_service.cc:
```diff
@@ -1876,33 +1876,56 @@ future<std::map<gms::inet_address, float>> storage_service::effective_ownership(
+                            });
+                        }  catch (...) {
+                            // In case ss.get_ranges_for_endpoint(keyspace_name, endpoint) is not found, just mark it as zero and continue
+                            (*final_ownership)[endpoint] = 0;
```
Where did this come from? Just because you had an exception, doesn't mean it's okay to return zero!

amnonh replied:
get_ranges_for_endpoint throws if there is no keyspace,endpoint match.
Which means that the specific endpoint has zero ownership.
Note that it only catches an exception that happens in get_ranges_for_endpoint.
I can catch only that specific exception to be on the safe side.

avikivity replied:
`...` is not a specific exception. Any bad_alloc (or some other exception) here will be caught too.

amnonh replied:
right, that's why I wrote that I'll change it to a specific exception so that only the case that there is no keyspace for the endpoint will be caught.

avikivity replied:
Ah, I somehow missed "can"

amnonh commented:
@avikivity I have forced-push a new version with the following changes:

* final_ownership is now in the do_with.
* removed unneeded explicit constructors from do_with.
* changed the catch to an explicit exception.

`get_ranges_for_endpoint` remains a loop, but it is inside `do_for_each` so the code will be preempted between calls if needed.

slivne commented:
@asias please review as well
